diffrax/_adjoint.py

```diff
@@ -852,3 +865,427 @@ def loop(
+        ```
+
+        ```bibtex
+        @article{zhuang2021mali,
+            title={Mali: A memory efficient and reverse accurate integrator for neural
+                    odes},
+            author={Zhuang, Juntang and Dvornek, Nicha C and Tatikonda, Sekhar and
+            Duncan, James S},
+            journal={arXiv preprint arXiv:2102.04668},
+            year={2021}
+        }
+        ```
+    """
+
+    l: float = 0.999
+
+    def loop(
+        self,
+        *,
+        args,
+        terms,
+        solver,
+        saveat,
+        init_state,
+        passed_solver_state,
+        passed_controller_state,
+        event,
+        **kwargs,
+    ):
+        # `is` check because this may return a Tracer from SaveAt(ts=<array>)
+        if (
+            eqx.tree_equal(saveat, SaveAt(t1=True)) is not True
+            and eqx.tree_equal(saveat, SaveAt(steps=True)) is not True
+            and eqx.tree_equal(saveat, SaveAt(t0=True, steps=True)) is not True
+        ):
+            raise ValueError(
+                "Can only use `diffrax.ReversibleAdjoint` with "
+                "`saveat=SaveAt(t1=True)` or `saveat=SaveAt(steps=True)`."
+            )
+
+        if event is not None:
+            raise NotImplementedError(
+                "`diffrax.ReversibleAdjoint` is not compatible with events."
+            )
+
+        if isinstance(solver, (SemiImplicitEuler, ReversibleHeun, LeapfrogMidpoint)):
+            raise ValueError(
+                "`diffrax.ReversibleAdjoint` is not compatible with solvers that are "
+                f"intrinsically algebraically reversible, such as {solver}."
+            )
+
+        if is_unsafe_sde(terms):
+            raise ValueError(
+                "`adjoint=ReversibleAdjoint()` does not support `UnsafeBrownianPath`. "
+                "Consider using `VirtualBrownianTree` instead."
+            )
+        if is_sde(terms):
+            if isinstance(solver, AbstractItoSolver):
+                raise NotImplementedError(
+                    f"`{solver.__class__.__name__}` converges to the Itô solution. "
+                    "However `ReversibleAdjoint` currently only supports Stratonovich "
+                    "SDEs."
+                )
+            elif not isinstance(solver, AbstractStratonovichSolver):
+                warnings.warn(
+                    f"{solver.__class__.__name__} is not marked as converging to "
+                    "either the Itô or the Stratonovich solution. Note that "
+                    "`ReversibleAdjoint` will only produce the correct solution for "
+                    "Stratonovich SDEs."
+                )
+
+        solver = _Reversible(solver, self.l)
+        tprev = init_state.tprev
+        tnext = init_state.tnext
+        y = init_state.y
+
+        init_state = eqx.tree_at(
+            lambda s: s.solver_state,
+            init_state,
+            solver.init(terms, tprev, tnext, y, args),
+            is_leaf=_is_none,
+        )
```

Comment:
Note that the `_Reversible.init` here will re-call the underlying original `solver.init`, I think unnecessarily? I think we should be able to do just `init_state = eqx.tree_at(lambda s: s.solver_state, init_state, (init_state.solver_state, init_state.y))`, and then set `class _Reversible: def init(...): assert False`.